overlord/restart/restart.go

```diff
@@ -157,6 +254,67 @@ func Request(st *state.State, t RestartType, rebootInfo *boot.RebootInfo) {
 	rm.handleRestart(t, rebootInfo)
 }
 
+func changeHeldForSystemRestart(chg *state.Change) (int, error) {
+	var count int
+	err := chg.Get("held-for-system-restart", &count)
+	if err != nil && !errors.Is(err, state.ErrNoState) {
+		return 0, err
+	}
+	return count, nil
+}
+
+func incrementHeldForSystemRestart(chg *state.Change) error {
+	if chg == nil {
+		// nothing to do
+		return nil
+	}
+	count, err := changeHeldForSystemRestart(chg)
+	if err != nil {
+		return err
+	}
+	chg.Set("held-for-system-restart", count+1)
+	return nil
+}
+
+// FinishTaskWithRestart will finish a task that needs a restart, by
+// setting its status and requesting a restart.
+// It should usually be invoked returning its result immediately
+// from the caller.
+// In some situations (classic) it might return state.Hold instead
+// to held progress until a non-automatic system restart occurs.
+// The restart manager itself will take of tracking that.
+func FinishTaskWithRestart(task *state.Task, status state.Status, rt RestartType, rebootInfo *boot.RebootInfo) error {
+	// if a system restart is requested on classic hold the task instead
+	// or just log the request if we are on the undo path
+	switch rt {
+	case RestartSystem, RestartSystemNow:
+		if release.OnClassic {
+			if status == state.DoneStatus {
+				rm := restartManager(task.State(), "internal error: cannot request a restart before RestartManager initialization")
+				// TODO notify the system
```

Comment:
Until we find a way to do this, should we use a snap warning?

Reply:
we should address this very soon afaiu